iio/iio_app/iio_app.h

```diff
@@ -71,21 +74,54 @@ struct iio_app_device {
 };
 
 /**
- * @brief Register devices and start an iio application
- *
- * Configuration for communication is done in parameters.h
- * @param ctx_attrs - array of context attribute name/value pairs.
- * @param nb_ctx_attr - number of context attributes in the array above.
- * @param devices - is an array of devices to register to iiod
- * @param len - is the number of devices
- * @return 0 on success, negative value otherwise
+ * @struct iio_app_desc
+ * @brief IIO application descriptor.
+ */
+struct iio_app_desc {
+	/** Array of context attribute name/value pairs */
+	struct iio_ctx_attr *ctx_attrs;
+	/** Number of context attributes in the array above */
+	uint32_t nb_ctx_attr;
+	struct iio_app_device *devices;
+	uint32_t nb_devices;
+	struct no_os_uart_init_param *user_uart_params;
+	/** IIO triggers to be used */
+	struct iio_trigger_init *trigs;
+	/** Number of triggers to be used */
+	int32_t nb_trigs;
+	/** IRQ descriptor to be used */
+	void *irq_desc;
+	/**  IIO descriptor to be returned */
+	struct iio_desc *iio_desc;
+	/** Function to be called each step */
+	int (*app_callback)(void *arg);
+	/** Function parameteres */
+	void *arg;
+};
+
+/**
+ * @struct iio_app_init_param
+ * @brief IIO application descriptor initialization parameters.
  */
-int32_t iio_app_run(struct iio_ctx_attr *ctx_attrs, uint32_t nb_ctx_attr,
-		    struct iio_app_device *devices, uint32_t len);
+struct iio_app_init_param {
+	struct iio_ctx_attr *ctx_attrs;
+	uint32_t nb_ctx_attr;
+	struct iio_app_device *devices;
+	uint32_t nb_devices;
+	struct iio_trigger_init *trigs;
+	int32_t nb_trigs;
+	struct no_os_uart_init_param user_uart_init_params;
+	void *irq_desc;
+	struct iio_desc *iio_desc;
+	int (*app_callback)(void *arg);
+	void *arg;
+};
+
+/** Register devices for an iio application */
+int32_t iio_app_init(struct iio_app_desc **app,
+	struct iio_app_init_param app_init_param);
+
+/** Start an iio application */
+int32_t iio_app_run(struct iio_app_desc *app);
```

Comment:
I'm wondering if we should add a remove() function too. It doesn't make much sense yet it looks weird to allocate memory without a function which deallocates it.

Reply:
Yes, we can.